codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java

```diff
@@ -543,15 +543,12 @@ private static int decodeMessageId(ByteBuf buffer) {
     private static Result<MqttSubAckPayload> decodeSubackPayload(
             ByteBuf buffer,
             int bytesRemainingInVariablePart) {
-        final List<Integer> grantedQos = new ArrayList<Integer>();
+        final List<Integer> grantedQos = new ArrayList<Integer>(bytesRemainingInVariablePart);
         int numberOfBytesConsumed = 0;
         while (numberOfBytesConsumed < bytesRemainingInVariablePart) {
-            int qos = buffer.readUnsignedByte();
-            if (qos != MqttQoS.FAILURE.value()) {
-                qos &= 0x03;
-            }
+            int reasonCode = buffer.readUnsignedByte();
```

Comment:
Do we need some special handling for MQTT 3.x ?

Reply:
No special handling - this just cuts off some codes which were not valid in 3.x but became valid in 5.0. See http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349807 vs https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901178.

Reply:
The question is do we need to cut these off when 3.x is used ?

Reply:
I don't see much use in it, and as checking the protocol version has some performance penalty (need to retrieve it from the channel attribute) - decided not to do it. If you think someone might need it - I can add the check.

Reply:
Ok I was just wondering what will happen to users that "depend" on this... So you think there is no "risk" ?

Reply:
With well-behaved MQTT server they'll see exactly the same behavior, as MQTT 3.1 requires bits othter than 0,1 and 7 to remain 0 when sending a message. They only will see unexpected code if server violates the spec, which is possible, in principle, but is very unlikely and, well, it's spec violation from the server side.